pfcpiface/p4rt_translator.go

```diff
@@ -0,0 +1,642 @@
+
+	TableInterfaces           = "PreQosPipe.interfaces"
+	TableTunnelPeers          = "PreQosPipe.tunnel_peers"
+	TableDownlinkTerminations = "PreQosPipe.terminations_downlink"
+	TableUplinkTerminations   = "PreQosPipe.terminations_uplink"
+	TableDownlinkSessions     = "PreQosPipe.sessions_downlink"
+	TableUplinkSessions       = "PreQosPipe.sessions_uplink"
+
+	ActSetUplinkSession       = "PreQosPipe.set_session_uplink"
+	ActSetDownlinkSession     = "PreQosPipe.set_session_downlink"
+	ActSetDownlinkSessionBuff = "PreQosPipe.set_session_downlink_buff"
+	ActUplinkTermDrop         = "PreQosPipe.uplink_term_drop"
+	ActUplinkTermFwd          = "PreQosPipe.uplink_term_fwd"
+	ActDownlinkTermDrop       = "PreQosPipe.downlink_term_drop"
+	ActDownlinkTermFwd        = "PreQosPipe.downlink_term_fwd"
+	ActLoadTunnelParams       = "PreQosPipe.load_tunnel_param"
+
+	DefaultPriority = 0
+)
+
+type tunnelParams struct {
+	tunnelIP4Src uint32
+	tunnelIP4Dst uint32
+	tunnelPort   uint16
+}
+
+type P4rtTranslator struct {
+	p4Info p4ConfigV1.P4Info
+}
+
+func newP4RtTranslator(p4info p4ConfigV1.P4Info) *P4rtTranslator {
+	return &P4rtTranslator{
+		p4Info: p4info,
+	}
+}
+
+func convertValueToBinary(value interface{}) ([]byte, error) {
+	switch t := value.(type) {
+	case []byte:
+		return value.([]byte), nil
+	case bool:
+		uintFlag := uint8(0)
+
+		flag := value.(bool)
+		if flag {
+			uintFlag = 1
+		}
+
+		b := make([]byte, 1)
+		b[0] = uintFlag
+
+		return b, nil
+	case uint8:
+		b := make([]byte, 1)
+		b[0] = value.(uint8)
+
+		return b, nil
+	case uint16:
+		b := make([]byte, 2)
+		binary.BigEndian.PutUint16(b, value.(uint16))
+
+		return b, nil
+	case uint32:
+		b := make([]byte, 4)
+		binary.BigEndian.PutUint32(b, value.(uint32))
+
+		return b, nil
+	case uint64:
+		b := make([]byte, 8)
+		binary.BigEndian.PutUint64(b, value.(uint64))
+
+		return b, nil
+	default:
+		return nil, ErrOperationFailedWithParam("convert type to byte array", "type", t)
+	}
+}
+
+func (t *P4rtTranslator) tableID(name string) uint32 {
+	for _, table := range t.p4Info.Tables {
+		if table.Preamble.Name == name {
+			return table.Preamble.Id
+		}
+	}
+
+	return invalidID
+}
+
+func (t *P4rtTranslator) actionID(name string) uint32 {
+	for _, action := range t.p4Info.Actions {
+		if action.Preamble.Name == name {
+			return action.Preamble.Id
+		}
+	}
+
+	return invalidID
+}
+
+func (t *P4rtTranslator) getActionByID(actionID uint32) (*p4ConfigV1.Action, error) {
+	for _, action := range t.p4Info.Actions {
+		if action.Preamble.Id == actionID {
+			return action, nil
+		}
+	}
+
+	return nil, ErrNotFoundWithParam("action", "ID", actionID)
+}
+
+func (t *P4rtTranslator) getTableByID(tableID uint32) (*p4ConfigV1.Table, error) {
+	for _, table := range t.p4Info.Tables {
+		if table.Preamble.Id == tableID {
+			return table, nil
+		}
+	}
+
+	return nil, ErrNotFoundWithParam("table", "ID", tableID)
+}
+
+func (t *P4rtTranslator) getTableIDByName(name string) (uint32, error) {
```

Comment:
Something to keep in mind for future simplifications: similarly to how we we do it in the fabric-tna's pipeconf, we can automatically generate Go constants based on the P4Info file, e.g., using [this](https://github.com/opennetworkinglab/tassen/blob/master/util/go-gen-p4-const.py) script.